rfcs/004 - Continuous Two-Way Synchronization RFC.md

```diff
@@ -0,0 +1,160 @@
+
+That last piece is particularly interesting because it's a way to help us accelerate our development velocity and potentially turn our current "low-level modules" into "mid-level modules" that could be pure SQL modules depending on these RPC functions, and the "high-level modules" could decide to depend on either of them, depending on what is easier.
+
+### Alternatives Considered
+
+#### Just don't implement any `code-` modules
+
+Basically, we can ~~stick our heads in the sand~~ decide to enforce management of all cloud resources in a way that these changes are represented in our database, but this makes our ability to co-exist with other IaC tools impossible, as well as integrating with any sort of CI system (without basically re-implementing our own CI system we then demand users use). And therefore this approach has been rejected.
+
+#### Implement `iasql_commit` as replacement for `iasql_apply`/`iasql_sync` only
+
+Stick to manual mode and just replace `apply` and `sync` with `iasql_commit`. This allows us to implement the `code-` modules and requires less work, but keeps the harder to explain syntax in the forefront, rather than being a power user tool. It also precludes our ability to support Enforce/Track behavior described above, which narrows our set of advantages.
+
+This may still be a point along-the-way in the conversion to the Full Two-Way Mode, though, but has been rejected as the stopping point.
+
+#### Implement Two-Way Mode with the transaction-like syntax only
+
+Basically just the first half of the proposal, stopping before the Enforce/Track section. This is probably also a point along-the-way, but has been rejected as the stopping point because the remaining features are simpler to implement and provide significant functionality.
+
+## Expected Semver Impact
+
+If we were post v1.0.0, this would be a major update.
+
+## Affected Components
+
+- The `iasql_functions` module
+- The core RPC scheduler
+- The entire test suite
+- Most of the documentation
+
+## Expected Timeline
+
+This will take a few weeks to get across the line. The major steps (with estimated times) are:
+
+1. Implement the `iasql_commit` function and port test suite to using it. (1 week)
+2. Implement pausable cron and transaction functionality, make `apply/sync` no-ops, and update the test suite again. (1 week)
+3. Implement Enforce/Track/WaitFor and add new tests for them (3-4 days)
```

Comment:
I think the Enforce/Track section will be more difficult to implement or the one that could give us more problems.

Reply:
That may be the case. I just figured the test suite changes for the first transition would take more "grunt work" effort while this would be potentially more complicated mentally but more contained?

Reply:
I would split this off into three different steps:
3. Implement WaitFor and add tests for them
4. Implement Enforce/Track per db
5. Implement granular Enforce/Track per table